File: 0010-longest-substring-without-repeating-characters/solution.go
package main

func lengthOfLongestSubstring(s string) int {
	characters := []rune(s)
	set := make(map[rune]bool)

	left := 0
	maxLength := 0

	for right := 0; right < len(characters); right++ {

		for set[characters[right]] {
			delete(set, characters[left])
			left++
		}

		set[characters[right]] = true

		currentLength := right - left + 1

		if currentLength > maxLength {
			maxLength = currentLength
		}
	}

	return maxLength
}

File: 0010-longest-substring-without-repeating-characters/solution.swift
func lengthOfLongestSubstring(_ s: String) -> Int {
    let characters = Array(s)
    var set = Set<Character>()

    var left = 0
    var maxLength = 0

    for right in 0..<characters.count {
        while set.contains(characters[right]) {
            set.remove(characters[left])
            left += 1
        }

        set.insert(characters[right])

        maxLength = max(maxLength, right - left + 1)
    }

    return maxLength
}

//Time  → O(n)
//Space → O(n)
